### ISSUE TITLE
[audio] Fails to initialize PulseAudio context

### ISSUE THREAD

**Issue by raysan5** [raysan5/raylib#575]
On Raspberry Pi platform, `mini_al` tries to open PulseAudio backend but it's not available and fails.

As per [this answer](https://raspberrypi.stackexchange.com/questions/639/how-to-get-pulseaudio-running), it seems PulseAudio on RPI should be enabled through ALSA driver.

Pointing to @mackron.

**Comment by mackron** [comment]
So audio is outright failing to do initialize and you do not get any audio output? The way mini_al should work is that it just seamlessly skips past PulseAudio (or any other unsupported backend) if it can't use it. If Pulse Audio is not installed it should fall back to ALSA.

I will test mini_al again on RPI, but I know I tested that before the latest release. Are you getting a crash?

**Comment by mackron** [comment]
I tested mini_al again on RPI and it's working the way I would expect (not a fan that latency, though!). The PulseAudio backend is the highest priority backend on Linux, and as such it attempts PulseAudio first. This will fail because PulseAudio is not installed. It then attempts ALSA in which case it succeeds. This is the intended chain of events for Linux platforms. If, however, it is not falling back to ALSA, or ALSA is also failing to initialize, that's another problem I'll need to look at.

If you don't like that it's attempting PulseAudio before ALSA, you can always the change the order of priority when initializing mini_al. I don't want to add distro-specific priorities to mini_al, though, because it just makes things too confusing.

**Comment by mackron** [comment]
A fix for this is in mini_al's dev branch, but I have a few other important bug fixes to do before doing a new release so I'll wait until those fixes are in before updating raylib. If you like you can take a copy of mini_al from the dev branch yourself if you need it right now.

Thanks for letting me know about this!

**Comment by raysan5** [comment]
Wow @mackron! What a quick and detailed response! Thank you very much for looking into this! :D

Current priority is perfect, I agree it's better to avoid distro-specific priorities.

I'm about to release next raylib version, so, I'll pick current `mini_al` from dev branch for now, you can just send further updates later on.

Thank you very much for your time an effort! :D

**Comment by raysan5** [comment]
Just tested it with audio_music_stream and despite audio example runs, it hangs apparently when buffer is consumed and requires refilling...

I'll keep testing it.

**Comment by mackron** [comment]
Only on RPI?

**Comment by raysan5** [comment]
@mackron yes, only on the RPI, it seems at refill point it gets stuck on an infinite loop because I loose control over the program and neither input is processed, so, I need to restart the RPI.

**Comment by mackron** [comment]
An emergency fix for now is to add `deviceConfig.alsa.noMMap = MAL_TRUE;` just before the call to `mal_device_init()`. I'm still investigating the root cause of this issue.

(MMAP mode with ALSA has been the most annoying thing to get right in the whole of mini_al. Grrr!)

**Comment by raysan5** [comment]
Hey @mackron! Thank you very much for the quick answer and fix! Excuse me for bothering you with those issues... I understand that RPI is not a priority platform... sorry.

I'll try those fixes as soon as possible. Thanks again for your hard work! :)

**Comment by mackron** [comment]
It's no problem. There's actually another problem I detected while testing this - there's glitches when running audio through my USB device which is due to the automatic buffer size being too small. I'll need to play around with that as well when I get the chance.
